dplock/usage.go

```diff
@@ -0,0 +1,107 @@
+package dplock
+
+import (
+	"time"
+)
+
+const TimeThresholdSinceLastRelease = 100 * time.Millisecond
+const UsageSleep = 50 * time.Millisecond
+const MaxCount = 10
+
+// Usages is a 2D map to keep track of lock usages by resourceID and owner
+type Usages map[string]map[string]*Usage
+
+// Usage keeps track of locks that have been acquired by a particular caller (lock owner) and resource
+type Usage struct {
+	Count    int       // counter for the number of times that a lock has been acquired on the first attempt within a short period of time after being released
+	Released time.Time // timestamp for the last time that a lock was released
+}
+
+// getUsage returns the Usage for the provided resourceID and owner, if it exists
+func (u Usages) getUsage(resourceID, owner string) (*Usage, bool) {
+	resUsages, found := u[resourceID]
+	if !found {
+		return nil, false
+	}
+	usage, found := resUsages[owner]
+	return usage, found
+}
+
+// SetCount increases the counter if the lock has been previously released in the last 'timeThresholdSinceLastRelease'
+// otherwise it resets the counter to 0
+// if the Usage did not exist in the map, it will be created
+func (u Usages) SetCount(resourceID, owner string) {
+	resUsages, found := u[resourceID]
+	if !found {
+		resUsages = map[string]*Usage{}
+		u[resourceID] = resUsages
+	}
+	usage, found := resUsages[owner]
+	if !found {
+		usage = &Usage{}
+		resUsages[owner] = usage
+	}
+	if usage.Released.IsZero() || time.Since(usage.Released) <= TimeThresholdSinceLastRelease {
+		usage.Count++ // increase count because the lock was released by the same caller a short period of time ago
+	} else {
+		usage.Count = 0 // reset count because the lock was released by the same caller a long period of time ago
+	}
+}
+
+// WaitIfNeeded sleeps for 'usageSleep' time if the provided resource has been locked by the provided owner at least MaxCount times,
+// with a period of time smaller than 'timeThresholdSinceLastRelease' between releasing and re-acquiring the lock for all times.
+// After sleeping, the counter is reset to 0
+func (u Usages) WaitIfNeeded(resourceID, owner string) {
+	usage, found := u.getUsage(resourceID, owner)
+	if !found {
+		return
+	}
+	if usage.Count >= MaxCount && time.Since(usage.Released) <= TimeThresholdSinceLastRelease {
+		Sleep(UsageSleep)
+		usage.Count = 0
+	}
+}
+
+var Sleep = func(d time.Duration) {
```

Comment:
I don't think this needs to be a separate function and could be inlined as it is only called from line 60

Reply:
I had to create a wrapper function to be able to create unit tests without sleeping, and validating the sleep time (i.e. we don't want unit tests to sleep, and we want them to check that the code called sleep instead).